libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp

```diff
@@ -1366,4 +1417,184 @@ void PeerConnection::GetStats(webrtc::RTCStatsCollectorCallback* callback) {
 PeerConnection::PeerConnection(RefPtr<GlobalFactory> global_factory)
     : TrackedObject(std::move(global_factory), ObjectType::kPeerConnection) {}
 
+void AudioReadStream::audioFrameCallback(const void* audio_data,
+                                         const uint32_t bits_per_sample,
+                                         const uint32_t sample_rate,
+                                         const uint32_t number_of_channels,
+                                         const uint32_t number_of_frames) {
+  std::lock_guard<std::mutex> lock(frames_mutex_);
+  // maintain buffering limits, after adding this frame
+  const size_t maxFrames = std::max(buffer_ms_ / 10, 1);
+  while (frames_.size() > maxFrames) {
+    frames_.pop_front();
+  }
+  // add the new frame
+  auto& frame = frames_.emplace_back();
+  frame.bits_per_sample = bits_per_sample;
+  frame.sample_rate = sample_rate;
+  frame.number_of_channels = number_of_channels;
+  frame.number_of_frames = number_of_frames;
+  size_t size =
+      (size_t)(bits_per_sample / 8) * number_of_channels * number_of_frames;
+  auto src_bytes = static_cast<const std::byte*>(audio_data);
+  frame.audio_data.insert(frame.audio_data.begin(), src_bytes,
+                          src_bytes + size);
+}
+
+void AudioReadStream::staticAudioFrameCallback(
+    void* user_data,
+    const AudioFrame& frame) {
+  auto ars = static_cast<AudioReadStream*>(user_data);
+  ars->audioFrameCallback(frame.data_, frame.bits_per_sample_, frame.sampling_rate_hz_,
+                          frame.channel_count_, frame.sample_count_);
+}
+
+AudioReadStream::AudioReadStream(PeerConnection* peer, int bufferMs)
+    : peer_(peer),
+      buffer_ms_(bufferMs >= 10 ? bufferMs : 500 /*TODO good value?*/) {
+  peer->RegisterRemoteAudioFrameCallback(
+      AudioFrameReadyCallback{&staticAudioFrameCallback, this});
+}
+
+AudioReadStream::~AudioReadStream() {
+  peer_->RegisterRemoteAudioFrameCallback(AudioFrameReadyCallback{});
+}
+
+AudioReadStream::Buffer::Buffer() {
+  resampler_ = std::make_unique<webrtc::Resampler>();
+}
+AudioReadStream::Buffer::~Buffer() {}
+
+void AudioReadStream::Buffer::addFrame(const Frame& frame,
```

Comment:
@stephenatwork IIRC you were saying that the first passes are probably dwarfed by resampling (and all of this by video processing)?

Anyway I agree there is a lot of space for optimization here (and that it doesn't need to be tackled now).

Reply:
Yes, these passes are a tiny fraction of frame time. The existing code is pretty dumb because it handles some unlikely paths.

One easy optimization to make is that for headsets, the most common transform is mono16 -> stereo f32.
So a dedicated code path for mono -> stereo could avoid some copies and half the resampling.

One other thing is that these simple loops tend to auto-vectorize very well if the compiler can be given enough hints (e.g. round buffer sizes/loop counts up to 8 or 16).

Reply:
Frames are going to be relatively small (960B for 10ms of 16bit 48KHz), so multiple passes shouldn't be an issue. I have reordered the code a bit so that we do the resampling on just one channel when possible, that should be good enough.